ts/src/profiler.ts

```diff
@@ -13,47 +13,335 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import * as path from 'path';
 
-import {AuthenticationConfig, Common, Logger, Service, ServiceConfig} from '../third_party/types/common-types';
+import * as path from 'path';
+import * as pify from 'pify';
+import * as zlib from 'zlib';
 
-import {Config, ProfilerConfig} from './config';
+import {perftools} from '../../proto/profile';
+import {Common, Logger, Service, ServiceObject} from '../third_party/types/common-types';
+import {ProfilerConfig} from './config';
 import {HeapProfiler} from './profilers/heap-profiler';
 import {TimeProfiler} from './profilers/time-profiler';
 
+export const common: Common = require('@google-cloud/common');
+const parseDuration: (str: string) => number = require('parse-duration');
 const pjson = require('../../package.json');
-const common: Common = require('@google-cloud/common');
+const API = 'https://cloudprofiler.googleapis.com/v2';
+const SCOPE = 'https://www.googleapis.com/auth/monitoring.write';
+const gzip = pify(zlib.gzip);
+
+enum ProfileTypes {
+  Wall = 'WALL',
+  Heap = 'HEAP'
+}
+
+/**
+ * @return - true if http status code indicates an error and false otherwise.
+ */
+function isErrorResponseStatusCode(code: number) {
+  return code < 200 || code >= 300;
+}
+
+/**
+ * Interface for deployment field of RequestProfile. Profiles with matching
+ * deployments will be grouped together.
+ * Used as body of request when creating profile using the profiler API.
+ *
+ * Public for testing.
+ */
+export interface Deployment {
+  projectId?: string;
+  target?: string;
+  labels?: {zone?: string, version?: string};
+}
+
+/**
+ * Interface for body of response from profiler API when creating
+ * profile and used as body of request to profiler API when
+ * uploading a profile.
+ *
+ * Public for testing.
+ */
+export interface RequestProfile {
```

Comment:
If we add a new field in the profile structure, will parsing fail? Will the new field be lost between us getting the profile from profile creation and sending it to UpdateProfile?

Reply:
This should not happen, and I added a test to createProfile() to confirm this.